controlplane/kubeadm/controllers/controller.go

```diff
@@ -580,3 +592,52 @@ func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, k
 
 	return nil
 }
+
+func (r *KubeadmControlPlaneReconciler) remediateUnhealthy(ctx context.Context, logger logr.Logger, controlPlane *internal.ControlPlane) error {
+	if !controlPlane.RemediationAllowed() || !controlPlane.HasUnhealthyMachine() || controlPlane.Machines.Len() < int(*controlPlane.KCP.Spec.Replicas) {
+		return nil
+	}
+
+	if controlPlane.HasDeletingMachine() {
+		return &capierrors.RequeueAfterError{RequeueAfter: deleteRequeueAfter}
+	}
+
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(controlPlane.Cluster))
+	if err != nil {
+		return errors.Wrap(err, "failed to create workload cluster client")
+	}
+
+	etcdStatus, err := workloadCluster.EtcdStatus(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch etcd status")
+	}
+
+	machine := controlPlane.UnhealthyMachines().Oldest()
+
+	if etcdStatus.FailureTolerance(machine) == 0 {
+		logger.Info("cluster has no failure tolerance, skipping remediation")
+		return nil
+	}
+
+	logger.Info("remediating unhealthy machine",
+		"machine", machine.Name,
+		"reason", conditions.GetReason(machine, clusterv1.MachineHealthCheckSuccededCondition),
+		"message", conditions.GetMessage(machine, clusterv1.MachineHealthCheckSuccededCondition),
+	)
+
+	if err := r.Client.Delete(ctx, machine); err != nil {
+		return errors.Wrap(err, "failed to delete unhealthy machine")
+	}
+
+	patchHelper, err := patch.NewHelper(machine, r.Client)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize patch helper")
+	}
+
+	conditions.MarkTrue(machine, clusterv1.MachineOwnerRemediatedCondition)
+	if err := patchHelper.Patch(ctx, machine); err != nil {
+		return errors.Wrap(err, "failed to patch unhealthy machine")
+	}
```

Comment:
Do we need to worry about possible issues if MachineOwnerRemediatedCondition is never set to True in the event that there is an error, crash, restart, etc between the deletion happening and the condition patch here?

What is the failure mode if that field is never set to True?

Is there any way we can add some type of reentrancy to ensure that we can go back and ensure it is set after a deletion is successful, but we may not have patched the condition?

Reply:
I think it's maybe fine, I don't think there's even a reason we need to do this other than it fulfills the contract we created, but as far as I know there is no code planned or intended that would observe a True `MachineOwnerRemediated` condition and do anything with it.

Reply:
@benmoss I'm good with ignoring for now if there are no existing/proposed workflows that would be impacted by it. It would probably be good to add something to the MHC doc around it as a potential future concern, though.